src/AutoMapper/PropertyMap.cs

```diff
@@ -146,151 +147,15 @@ internal void Seal(TypeMapRegistry typeMapRegistry)
-                getter = Expression.Default(_typeMap.DestinationType);
-            }
-            else
-            {
-                getter = destMember;
-            }
-
-            var destValueExpr = UseDestinationValue
-                ? getter
-                : Expression.Constant(null, DestinationPropertyType);
-
-            if (SourceType == null 
-                || (SourceType.IsEnumerableType() && SourceType != typeof (string)) 
-                || typeMapRegistry.GetTypeMap(new TypePair(SourceType, DestinationPropertyType)) != null 
-                || ((!EnumMapper.EnumToEnumMapping(new TypePair(SourceType, DestinationPropertyType)) ||
-                  EnumMapper.EnumToNullableTypeMapping(new TypePair(SourceType, DestinationPropertyType))) &&
-                 EnumMapper.EnumToEnumMapping(new TypePair(SourceType, DestinationPropertyType)))
-                || !DestinationPropertyType.IsAssignableFrom(SourceType))
-            {
-                //valueResolverExpr = 
-                //    (mappedObject, context) =>
-                //    {
-                //        var source = original(mappedObject, context);
-
-                //        return context.Mapper.Map(source, GetDestinationValue(mappedObject),
-                //            source?.GetType() ?? SourceType ?? context.SourceType,
-                //            DestinationPropertyType, context);
-                //    };
-                var sourceExpr = Expression.Parameter(typeof (object), "source");
-
-                var first = Expression.Assign(sourceExpr, valueResolverExpr.ToObject());
-                var ifTrue = SourceType != null 
-                    ? (Expression) Expression.Constant(SourceType, typeof(Type)) 
-                    : Expression.MakeMemberAccess(ctxtParam, typeof(ResolutionContext).GetProperty("SourceType"));
-                var ifFalse = Expression.Call(sourceExpr, typeof(object).GetMethod("GetType"));
-
-                var mapperProp = Expression.MakeMemberAccess(ctxtParam, typeof (ResolutionContext).GetProperty("Mapper"));
-                var mapMethod = typeof (IRuntimeMapper).GetMethod("Map", new[] {typeof (object), typeof (object), typeof (Type), typeof (Type), typeof (ResolutionContext)});
-                var second = Expression.Call(
-                    mapperProp,
-                    mapMethod,
-                    sourceExpr,
-                    destValueExpr.ToObject(),
-                    Expression.Condition(Expression.Equal(sourceExpr, Expression.Constant(null)), 
-                        ifTrue,
-                        ifFalse),
-                    Expression.Constant(DestinationPropertyType),
-                    ctxtParam
-                    );
-                valueResolverExpr = Expression.Block(typeof(object), new[] {sourceExpr}, first, second);
-            }
+            var lambda = this.CreateExpression(typeMapRegistry);
+            var exp = lambda.ReplaceParameters(source, destination);

```

Comment:
Replace type parameters with conversions of objects.  Later will replace at the TypeMap level with the source and destination values allready converted once, instead of converting them for each type map.